flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java

```diff
@@ -0,0 +1,357 @@
+
+			// verify that we only issued a single request to the gateway
+			verify(jobManagerGateway, times(1)).requestJob(eq(jobId), any(Time.class));
+		}
+	}
+
+	/**
+	 * Tests that an AccessExecutionGraph is invalidated after its TTL expired.
+	 */
+	@Test
+	public void testExecutionGraphEntryInvalidation() throws Exception {
+		final Time timeout = Time.milliseconds(100L);
+		final Time timeToLive = Time.milliseconds(1L);
+		final JobID jobId = new JobID();
+		final AccessExecutionGraph accessExecutionGraph = mock(AccessExecutionGraph.class);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		when(jobManagerGateway.requestJob(eq(jobId), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph));
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture.get());
+
+			// sleep for the TTL
+			Thread.sleep(timeToLive.toMilliseconds());
+
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture2 = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture2.get());
+
+			verify(jobManagerGateway, times(2)).requestJob(eq(jobId), any(Time.class));
+		}
+	}
+
+
+	/**
+	 * Tests that a failure in requesting an AccessExecutionGraph from the gateway, will not create
+	 * a cache entry --> another cache request will trigger a new gateway request.
+	 */
+	@Test
+	public void testImmediateCacheInvalidationAfterFailure() throws Exception {
+		final Time timeout = Time.milliseconds(100L);
+		final Time timeToLive = Time.hours(1L);
+		final JobID jobId = new JobID();
+
+		final AccessExecutionGraph accessExecutionGraph = mock(AccessExecutionGraph.class);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		// let's first answer with a JobNotFoundException and then only with the correct result
+		when(jobManagerGateway.requestJob(eq(jobId), any(Time.class))).thenReturn(
+			FutureUtils.completedExceptionally(new JobNotFoundException(jobId)),
+			CompletableFuture.completedFuture(accessExecutionGraph));
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			try {
+				executionGraphFuture.get();
+
+				fail("The execution graph future should have been completed exceptionally.");
+			} catch (ExecutionException ee) {
+				assertTrue(ee.getCause() instanceof FlinkException);
+			}
+
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture2 = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture2.get());
```

Comment:
Maybe we should also verify that the gateway is requested two times `requestJob` here also, to check that a cache entry was not created for the first failing request.

Reply:
The former is checked implicitly by returning the `accessExecutionGraph`.

Not sure if we should test the component with such a level of detail. It basically involves a lot of implementation knowledge. From the user perspective it should not make a difference if we create a failed entry which is updated or no entry. The important bit is that it will retry it.

Reply:
I see. Ok, lets leave it as is then.